feat: add Context Menu to Runs view (ET-232)

johnkim-det opened the pull request:
<!---
## PR TITLE (Commit Body)
When squash-merging, GitHub will use this as the commit message.
Check the "Example Commit Body" for conventional commit semantics.
-->
## Ticket
<!---
A reference to the Jira ticket or Github issue. e.g. "[DET-1234]" or #123.
-->
ET-232


## Description
<!---
A description of the PR. For breaking changes, lead with "BREAKING CHANGE:".
--->
Adds context menu (menu which appears on right-click of DataGrid cell) to Runs view.

This PR includes these actions in the menu:

- Copy Value
- Open in New Tab
- Open in New Window
- Archive
- Unarchive
- Delete
- Kill
- Move

This PR has changes from https://github.com/determined-ai/determined/pull/9368 merged in. Planning not to merge this PR until that one is merged first. Should not actually include backend changes at that time. Code coverage numbers should hopefully also improve when the patch is only for context menu changes.

Most likely easier to review by looking at the diff between those two branches here: https://github.com/determined-ai/determined/compare/feature/actions-flat-run-table...ET-232

## Test Plan
<!---
Describe the scenarios in which you've tested your change, with screenshots as
appropriate. Reviewers may ask questions about this test plan to ensure adequate
coverage of changes.
-->
- Cells with a value should show a functional "Copy Value" option
- Cells with a link should show functional "Open in New Tab" and "Open in New Window" options

With permissions to modify, move, and delete experiments/runs:
- Cells should have functional "Move"/"Delete" options
- Cells for a run with killable state (for example, "Active" or "Paused") should show a functional "Kill" option
- Cells for an unarchived run should show functional "Archive" option
- Cells for an archived run should show functional "Unarchive" option



## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

<!---
Example Commit Body:
docs: tweak recommended "pip install" usage [DET-123]

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:

- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:

- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:

- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

johnkim-det commented:
Suggest reviewing diff at https://github.com/determined-ai/determined/compare/feature/actions-flat-run-table...ET-232, see Description above for explanation

keita-determined commented on webui/react/src/components/RunActionDropdown.tsx:
```diff
@@ -0,0 +1,190 @@
+}
+
+const Action = {
+  Copy: 'Copy Value',
```
bug?: when i copy value of `User`, `userId is copied`. ideally username (displayname?) should be copied. or avoid copying user value?

keita-determined commented on webui/react/src/components/RunActionDropdown.tsx:
```diff
@@ -0,0 +1,190 @@
+
+const Action = {
+  Copy: 'Copy Value',
+  NewTab: 'Open Link in New Tab',
+  NewWindow: 'Open Link in New Window',
```
qq: what is the condition to see these options?  its not visible somehow.

<img width="887" alt="image" src="https://github.com/determined-ai/determined/assets/109113616/a74d748f-bf5c-4bf1-9ce5-419e1dd0403f">

johnkim-det replied:
needs to be a link (id, name)

keita-determined replied:
oh yes. that was a dumb question.

keita-determined commented on webui/react/src/components/RunActionDropdown.tsx:
```diff
@@ -0,0 +1,190 @@
+  return (
+    <>
+      <Dropdown isContextMenu menu={dropdownMenu} open={makeOpen} onClick={handleDropdown}>
+        <div />
```
is this neeeded?

johnkim-det replied:
we do need it to prevent error in `Dropdown` component -- could probably refactor the Dropdown component to make it not required but I think that would be a separate issue.

keita-determined replied:
interesting

keita-determined commented on webui/react/src/components/RunActionDropdown.tsx:
```diff
@@ -0,0 +1,190 @@
+  );
+
+  const shared = (
+    <>
```
super nit: i think we can get rid of fragment

keita-determined commented on webui/react/src/pages/FlatRuns/FlatRuns.tsx:
```diff
@@ -974,6 +1000,30 @@ const FlatRuns: React.FC<Props> = ({ projectId, searchId }) => {
+                  onComplete={onComplete}
+                  onLink={onClose}
+                  onVisibleChange={onVisibleChange}>
+                  {/* <div /> */}
```
```suggestion
```

keita-determined commented on webui/react/src/components/RunActionDropdown.tsx:
```diff
@@ -0,0 +1,190 @@
+  filterFormSetWithoutId,
+  projectId,
+}: Props) => {
+  const id = run.id;
```
can we directly use `run.id` instead of `id`? i feel like id sounds like a too common word in this component because there are run id, project id etc. i think explicitly using `run.id` is better to understand what the `id` is

keita-determined reviewed (commented):
the test coverage of this PR is slightly lower than 80%. can we add a bit more test cases?

---
its hard to tell if the action is completed or not. can we add notifications? 

https://github.com/determined-ai/determined/assets/109113616/4f5c65db-bafe-4413-9012-4f60bcb7b864

johnkim-det commented:
@keita-determined addressed feedback except for the codecov issue -- I think coverage numbers will be different when it's compared to main with your PR merged.

keita-determined reviewed (commented):
i'll re-review this PR when the action PR is merged, but so far LGTM

keita-determined commented:
can we please add more test cases? the coverage is 67.66055%

johnkim-det commented:
@keita-determined updated w more tests

keita-determined commented on webui/react/src/components/ExperimentActionDropdown.test.tsx:
can we fix the following warning?

```
Warning: An update to ForwardRef inside a test was not wrapped in act(...).

When testing, code that causes React state updates should be wrapped into act(...):

act(() => {
  /* fire events that update state */
});
/* assert on the output */
```

keita-determined reviewed (approved):
added few comments, otherwise \o/
